src/builder/context.rs

```diff
@@ -112,26 +117,102 @@ impl<'a> Context<'a> {
-            }
-            let path = Path::new("/vagga/root").join(path);
-            try_msg!(Dir::new(&path).recursive(true).create(),
-                 "Error creating cache dir: {err}");
-            clean_dir(&path, false)?;
-            try_msg!(BindMount::new(&cache_dir, &path).mount(),
-                "mount cache dir: {err}");
-            self.mounted.push(path);
+        let path = Path::new("/vagga/root").join(path);
+        try_msg!(
+            Dir::new(&path).recursive(true).create(),
+            "Error creating cache dir: {err}"
+        );
+        clean_dir(&path, false)?;
+
+        try_msg!(
+            BindMount::new(&cache_dir, &path).mount(),
+            "mount cache dir: {err}"
+        );
+
+        self.mounted.push(path);
+        entry.insert(CachedDir { src: cache_dir, overlay_dir: None });
+        Ok(())
+    }
+
+    pub fn add_cache_dir_overlay(&mut self, path: &Path, name: &str)
+        -> Result<(), String>
+    {
+        let entry = match self.cache_dirs.entry(path.to_path_buf()) {
+            Entry::Vacant(entry) => entry,
+            Entry::Occupied(_) => return Ok(()),
+        };
+
+        let overlay_dir = Path::new("/vagga/container").join(name);
+        let work_dir = overlay_dir.join("work");
+        let upper_dir = overlay_dir.join("upper");
+        try_msg!(
+            Dir::new(&work_dir).recursive(true).create(),
+            "Error creating cache work dir: {err}"
+        );
+        try_msg!(
+            Dir::new(&upper_dir).recursive(true).create(),
+            "Error creating cache upper dir: {err}"
+        );
+
+        let cache_dir = Path::new("/vagga/cache").join(&name);
+        if !cache_dir.exists() {
+            try_msg!(
+                Dir::new(&cache_dir).create(),
+                "Error creating cache dir: {err}"
+            );
         }
-        return Ok(());
+
+        let path = path.strip_prefix("/")
+            .map_err(|_| format!("cache_dir must be absolute: {:?}", path))?;
+        let path = Path::new("/vagga/root").join(path);
+        try_msg!(
+            Dir::new(&path).recursive(true).create(),
+            "Error creating cache dir: {err}"
+        );
+        clean_dir(&path, false)?;
+
+        try_msg!(
+            Overlay::writable(
+                vec!(cache_dir.as_path()).into_iter(),
+                &upper_dir,
+                &work_dir,
+                path.as_path(),
+            )
+            .mount(),
+            "mount cache dir: {err}"
+        );
+
+        self.mounted.push(path);
+        entry.insert(CachedDir { src: cache_dir, overlay_dir: Some(overlay_dir) });
+        Ok(())
+    }
+
+    pub fn get_cached_dir(&self, path: &Path) -> Option<&CachedDir> {
+        self.cache_dirs.get(path)
+    }
+
+    pub fn is_cached_dir_overlay(&self, path: &Path) -> bool {
+        self.get_cached_dir(path)
+            .and_then(|d| d.overlay_dir.as_ref())
+            .is_none()
```

Comment:
I'm a bit confused. Should this be `is_some`?

Reply:
Me too )))